### PR TITLE
14 return single values if owl indicates property maximum cardinality is 1

### PR DESCRIPTION
**Pull request by OwenR-York**
**Changes to the model loading/processing and related UI.**

- RDFModels are now loaded using Jena's OntModel which adds some classes and properties to a simple RDF Model.
- Schemas URIs can be added to the model editor dialogue.
- Data models and schemas are automatically loaded using a reasoner with Jena's default OWL settings.
- Property values are now limited by schema-inferred restrictions (max cardinality)

**Refactoring**

- RDFPropertyProcesses.java has been added to provide some supporting state-less processing for RDFResource.getProperty().
- Code has been refactored from RDFResource.getProperty() to RDFPropertyProcesses, this reduces the complexity of the method .getProperty()

**Test and examples**

- OWL demo data contains a schema and model (which includes Max Cardinality)
- The original tests have been updated to account for the changes to the RDF model that occur due to Jena's OntModel representations being used to store an RDF Model. (Notably, Jena added several OWL classes/properties to a basic RDF model, which broke several tests that checked for a specific number of classes, etc.).
- Some tests have been applied using the OWL demo data and schema, these check counts of classes/properties/restrictions that should not change unless another schema or reasoner is used.
- The max cardinality limitations on properties are tested, a query for a property should return 2 values by max cardinality reduces this to 1.
- Exceeding max cardinality when querying a property value results in a System.err reporting that the returned property values have been pruned.

**Future work/note**

- Add more refined control for selecting Jena's Reasoner.
- Loading a data model into an OntModel adds information to a model which may not be desirable for output back to a user's data model file. When enabling editing and saving of user data models, a user's data model may need to be loaded as a plain RDFModel type. Further investigation into Jena's API around the OntModel might reveal options to write out models that commit the additional OWL and schema information.

### CODE REVIEW

File: bundles/org.eclipse.epsilon.emc.rdf/src/org/eclipse/epsilon/emc/rdf/RDFModel.java [hunk 1]
```diff
@@ -96,7 +137,39 @@ public RDFResource getElementById(String iri) {
 		}
 		return null;
 	}
+
```
Comment (agarciadom) [inline bundles/org.eclipse.epsilon.emc.rdf/src/org/eclipse/epsilon/emc/rdf/RDFModel.java:140]: If users are meant to use these methods (e.g. via `Model.listClassesForElementsByID`, then they should be documented in the README).

I'd suggest renaming them from `list*` to `get*`, because in that case the default reflection-based property getter in Epsilon should be able to resolve things like `Model.ontClasses`. I would also drop the `InModel` suffix as it's redundant - you are calling an `RDFModel` method, we already know you're asking something at the model level.

Comment (agarciadom) [reply bundles/org.eclipse.epsilon.emc.rdf/src/org/eclipse/epsilon/emc/rdf/RDFModel.java:140]: I would also use less technical-sounding names, like:

* `Model.getClassesForElementByID(iri, direct)`. You need to document what `direct` means.
* `Model.getOntologyClasses()` (which would be also available as `Model.ontologyClasses`).
* `Model.getOntologyProperties()` (also available as `Model.ontologyProperties`).
* `Model.getRestrictions()` (also available as `Model.restrictions`).

**Comment by agarciadom** [comment]
I've had a look at the PR. It's almost there, but it doesn't quite go all the way: it's missing a test with the actual property getter ensuring that `element.hasMotherBoard` returns a single element, as in, the actual object and not a collection.

Remember that the idea is that we could reuse the same Epsilon script as in the original EMF metamodel (prior to MOF2RDF). If we had an EMF `EClass` with `EReference hasMotherBoard`, `x.hasMotherBoard` would just give you the motherboard or null, and not a collection.

I'll add a commit to do this now.

**Comment by agarciadom** [comment]
I have changed the logic and the tests to what we want: return the actual value if max cardinality is 1, instead of a collection.

We still have some things to do: see my comment above about the `list*InModel` methods needing a rename and to be documented in the README, and I have also added a few `TODO`s to the `RDFModelOWLReasonerTest` class, which needs a few more tests (at least one for the warning, and one for returning `null` from `x.hasMotherBoard`).

**Comment by agarciadom** [comment]
Another minor comment: in the RDF Model configuration dialog, I'd personally put the schema URLs above the data URLs, as the schema URLs in a way are at a "higher level" than the data ones.

We can ask Dimitris and Gian what they think, though. This is just a matter of taste :-).

**Comment by agarciadom** [comment]
To avoid confusion, we should rename hasMotherboard to motherboard in the example.

**Comment by agarciadom** [comment]
Using an invalid RDF model as an example could create confusion in users. I'd change this model to be valid (removing the extra motherboards), and I'd only introduce an invalid model as an example of how the Jena-based validation works when we work in #20. In any case, Gian commented that the Semantic Web community doesn't like to abuse OWL for validation.

I'd also try what happens in the example Gian gave of an RDF model where `Animal`s have `legs`, and we classify  `Animal`s into Monoped, Biped and Quadruped based on the number of legs they have. Try doing this:

```
Monoped.all.println();
Biped.all.println();
Quadruped.all.println();

for (a in Animal.all) {
  a.legs.println('Animal ' + a + ' legs: ');
}
```

This actually creates an interesting scenario, where we may be getting a single value from `Monoped` (since there will be a restriction that they have one leg), but in theory we'd still want a list.

(If you need an example of a Monoped, we could try [limpets](https://en.wikipedia.org/wiki/Limpet).)

File: bundles/org.eclipse.epsilon.emc.rdf/src/org/eclipse/epsilon/emc/rdf/RDFModel.java (hunk 1 above)
Comment (OwenR-York) [reply bundles/org.eclipse.epsilon.emc.rdf/src/org/eclipse/epsilon/emc/rdf/RDFModel.java]: After a conversation on how the EOL language works in terms of exposing public methods from this class for EOL users to call, these methods have been removed. They had been added as an example of some of the Jena API features that provide easy access to Classes/Properties/Restrictions.

**Comment by agarciadom** [comment]
OK, I think it's ready to merge now. I'll do a squash merge as there has been quite a bit of experimentation which doesn't need to be in the long-term history :-).
